Adds rake tasks to audit the Fedora 3 to Fedora 4 migration

hectorcorrea opened the pull request:
This pull request adds a few rake tasks to audit the migration of objects from Fedora 3 to Fedora 4. The idea is that these tasks will be run after deploying the Fedora 4 code base to production and immediately after the migration of data has been performed.

The code to perform the audit issues straight HTTP requests to Fedora 3 and Fedora 4 and bypasses Rubydora and ActiveFedora to make sure we query the Fedora repo as close to the metal as possible. 

The following Wiki page provides some information on how these rake tasks will be run in production: https://github.com/psu-stewardship/scholarsphere/wiki/Fedora-3-to-Fedora-4-Migration

hectorcorrea commented:
@awead @cam156 @mjgiarlo This is the latest on the code to audit the Fedora 3 to Fedora 4 migration. 

This code includes the suggestions that Adam made earlier this week (on a PR that I deleted.) Notice that I only added unit tests to test the code from the Fedora 4 point of view since I didn't want to be running tests against both Fedora 3 and Fedora 4 in the same code base. 

The unit tests stub the results that the Fedora 3 repo will have and then creates them in Fedora 4 and compares the two.

awead commented on app/models/migrate_audit.rb:
```diff
@@ -0,0 +1,30 @@
+
+  # This method populates the migrate_audit table with data 
+  # from the Fedora 3 repo.
+  def self.f3_audit!(fedora_url, fedora_user, fedora_password, namespace)
```
If looks like you're using these four arguments just to create `MigrateAuditFedora3`. Why not:
```
f3_audit!(MigrateAuditFedora3.new(fedora_url, fedora_user, fedora_password, namespace))
```
where `f3_audit!` is:
```ruby
def self.f3_audit! auditor = nil
  rails StandardError unless auditor.kind_of?(MigrateAuditFedora3)
  ...
```

hectorcorrea replied:
Cool. I like your approach better. Done.

awead commented on app/models/migrate_audit.rb:
```diff
@@ -0,0 +1,30 @@
+
+    audit = MigrateAuditFedora3.new(fedora_url, fedora_user, fedora_password, namespace)
+    f3_pids = audit.get_pids()
+    f3_pids.each do |pid|
```
consider:
`audit.get_pids.each do |pid|`

hectorcorrea replied:
Done.

awead commented on app/models/migrate_audit.rb:
```diff
@@ -0,0 +1,30 @@
+  # This method makes sure each record in the migrate_audit table is
+  # in the Fedora 4 repo and it has the same model as its Fedora 3
+  # counterpart.
+  def self.f4_audit(fedora_url, fedora_user, fedora_password)
```
see my previous comment. The args are only relevant to a MigrateAuditFedora4 object.

hectorcorrea replied:
Done.

awead commented on app/models/migrate_audit.rb:
```diff
@@ -0,0 +1,30 @@
+    # Run the audit.
+    f3_data = all
+    audit = MigrateAuditFedora4.new(fedora_url, fedora_user, fedora_password)
+    audit.audit(f3_data) do |result|
```
consider:
`audit.audit(all) do |result|`
and if you just passed a MigrateAuditFedora4 object, this line could read:
`auditor.audit(all) do |result|`
which reads a bit better.

hectorcorrea replied:
Indeed. Done.

awead commented on app/models/migrate_audit_fedora3.rb:
```diff
@@ -0,0 +1,130 @@
+  end
+
+  # Get the list of all PIDs in the Fedora 3 repository. 
+  def get_pids()
```
`()` are unnecessary in Ruby.

hectorcorrea replied:
They are now gone!

awead commented on app/models/migrate_audit_fedora3.rb:
```diff
@@ -0,0 +1,130 @@
+      #     <objectFields>
+      #     more objectFields
+      #   </resultList>
+      # <result>
```
This is documentation code and style-wise, it should be above the method call. See https://github.com/rdoc/rdoc for style pointers.

hectorcorrea replied:
Moved the comment to the top.

awead commented on app/models/migrate_audit_fedora3.rb:
```diff
@@ -0,0 +1,130 @@
+
+  # Get the list of all PIDs in the Fedora 3 repository. 
+  def get_pids()
+    batch_size = 50
```
Doesn't batch size always have to be 50? Why not make it an `attr_accessor` for flexibility.

hectorcorrea replied:
I missed this. Future enhancement if somebody feels strongly about it :)

awead commented on app/models/migrate_audit_fedora3.rb:
```diff
@@ -0,0 +1,130 @@
+    session_token = nil
+    while true
+      session = (session_token == nil) ? "" : "sessionToken=#{session_token}"
+      query_string = "query=&resultFormat=xml&maxResults=#{batch_size}&pid=true&#{session}"
```
Make `query_string` a method and you can simplify the session and session_token logic.
```
def query_string
  if session_token.nil?
    "query=&resultFormat=xml&maxResults=#{batch_size}&pid=true"
  else
    "query=&resultFormat=xml&maxResults=#{batch_size}&pid=true&#{session_token}"
  end
end
```
make an attr_accessor for session token and then `extract_session` is called directly to set the session_token value.

awead commented on app/models/migrate_audit_fedora3.rb:
```diff
@@ -0,0 +1,130 @@
+      #   </resultList>
+      # <result>
+      fedora_response = fedora_get("/objects", query_string)
+      xml = Nokogiri::XML(fedora_response)
```
why not:
`Nokogiri::XML(fedora_get("/objects", query_string))`

awead commented on app/models/migrate_audit_fedora3.rb:
```diff
@@ -0,0 +1,130 @@
+        all_pids.push pid 
+      end
+      session_token = extract_session_token(xml)
+      break if session_token == nil
```
break if session_token.nil?

hectorcorrea replied:
Done.

awead commented on app/models/migrate_audit_fedora3.rb:
```diff
@@ -0,0 +1,130 @@
+    #       </foxml:xmlContent>
+    #     </foxml:datastreamVersion>
+    #   </foxml:datastream>
+    #   ...
```
put documentation above the method call.

hectorcorrea replied:
Done.

awead commented on app/models/migrate_audit_fedora3.rb:
```diff
@@ -0,0 +1,130 @@
+    xml = Nokogiri::XML(fedora_response)
+    title = extract_title(xml)
+    has_model = extract_model(xml)
+    {pid: pid, has_model: has_model, title: title}
```
could condense this to:
```
xml = Nokogiri::XML(fedora_get("/objects/#{pid}/objectXML"))
{pid: pid, has_model: extract_model(xml), title: extract_title(xml)}

awead commented on app/models/migrate_audit_fedora3.rb:
```diff
@@ -0,0 +1,130 @@
+  def fedora_get(url_path, url_querystring = "")
+    uri = URI.parse(@fedora_url + url_path + "?" + url_querystring)
+    response = Net::HTTP.start(uri.hostname, uri.port) { |http|
+      request = Net::HTTP::Get.new(uri.path + "?" + url_querystring)
```
Not sure what the Ruby convention is, but I've seen `{}` for one-liners and `do ... end` for multiline stuff. @mjgiarlo any comment on that?

mjgiarlo replied:
Yup, that's exactly the convention we've been following, which is more or less this: https://github.com/bbatsov/ruby-style-guide

hectorcorrea replied:
Good catch. Fixed.

awead commented on app/models/migrate_audit_fedora3.rb:
```diff
@@ -0,0 +1,130 @@
+  private
+
+  def fedora_get(url_path, url_querystring = "")
+    uri = URI.parse(@fedora_url + url_path + "?" + url_querystring)
```
consider using `#{}` syntax for string interpolation as you did above.

awead commented on app/models/migrate_audit_fedora3.rb:
```diff
@@ -0,0 +1,130 @@
+
+  def extract_pids(xml)
+    docs = xml.xpath("//doc:pid", "doc" => "http://www.fedora.info/definitions/1/0/types/").children
+    docs.each do |doc|
```
consolidate to `children.each`

awead commented on app/models/migrate_audit_fedora3.rb:
```diff
@@ -0,0 +1,130 @@
+  def extract_pids(xml)
+    docs = xml.xpath("//doc:pid", "doc" => "http://www.fedora.info/definitions/1/0/types/").children
+    docs.each do |doc|
+      if doc.to_s.start_with?(@namespace + ":")
```
I think starting with the namespace is enough. You don't need the colon... maybe?

awead commented on app/models/migrate_audit_fedora3.rb:
```diff
@@ -0,0 +1,130 @@
+    docs = xml.xpath("//doc:pid", "doc" => "http://www.fedora.info/definitions/1/0/types/").children
+    docs.each do |doc|
+      if doc.to_s.start_with?(@namespace + ":")
+        yield doc.text
```
consolidate to:
`yield doc.text if doc.to_s.start_with ...`

awead commented on app/models/migrate_audit_fedora3.rb:
```diff
@@ -0,0 +1,130 @@
+    # of PIDs in the next execution of the query.
+    header = xml.xpath("//header:token", "header" => "http://www.fedora.info/definitions/1/0/types/").children
+    if header.length == 1
+      return header[0].text
```
consolidate:
`return ... if ...`

awead commented on app/models/migrate_audit_fedora3.rb:
```diff
@@ -0,0 +1,130 @@
+
+  def extract_session_token(xml)
+    # Get the session token. This is what allows us to fetch the next batch 
+    # of PIDs in the next execution of the query.
```
Ruby style usually puts the comments and explanation above the method call.

awead commented on app/models/migrate_audit_fedora3.rb:
```diff
@@ -0,0 +1,130 @@
+  end
+
+  def extract_title(xml)
+    # Extract the title from the DC data stream
```
comments about the method go above the method call.

awead commented on app/models/migrate_audit_fedora3.rb:
```diff
@@ -0,0 +1,130 @@
+  def extract_title(xml)
+    # Extract the title from the DC data stream
+    xml_set = xml.xpath("//dc:title", "dc" => "http://purl.org/dc/elements/1.1/")
+    unless xml_set.empty?
```
`xml.xpath("//dc:title", "dc" => "http://purl.org/dc/elements/1.1/").each` makes the unless block unnecessary.

awead commented on app/models/migrate_audit_fedora3.rb:
```diff
@@ -0,0 +1,130 @@
+    xml_set = xml.xpath("//dc:title", "dc" => "http://purl.org/dc/elements/1.1/")
+    unless xml_set.empty?
+      if xml_set.first.children.count > 0
+        return xml_set.first.children[0].text
```
`return ... if ...`

awead commented on app/models/migrate_audit_fedora3.rb:
```diff
@@ -0,0 +1,130 @@
+    # Extract the "has_model" (GenericFile, Batch, Collection) from the RELS-EXT data stream
+    xml_set = xml.xpath("//ns1:hasModel", "ns1" => "info:fedora/fedora-system:def/model#" )
+    unless xml_set.empty?
+      xml_node = xml_set.first
```
`xml.xpath("//ns1:hasModel", "ns1" => "info:fedora/fedora-system:def/model#" ).first` would be nil or the thing you want.

awead commented on app/models/migrate_audit_fedora3.rb:
```diff
@@ -0,0 +1,130 @@
+      xml_node = xml_set.first
+      if xml_node.attributes.key?("resource")
+        return xml_node.attributes["resource"].value
+      end
```
```
xml_node = xml.xpath("//ns1:hasModel", "ns1" => "info:fedora/fedora-system:def/model#" ).first
unless xml_node.nil?
  return xml_node.attributes["resource"].value if xml_node.attributes.key?("resource")
end
```
but the if statement may be unnecessary and NG would just return nil, so
```
return xml_node.attributes["resource"].value unless xml_node.nil?
```

awead commented on app/models/migrate_audit_fedora4.rb:
```diff
@@ -0,0 +1,97 @@
+      result.status = "Models mismatch. Expected: #{f3_model} but found: #{f4_model}" 
+    end
+    result
+  end
```
There's a lot to tease out here. You could opt to create a new Auditor class. Here's what I came up with:

```ruby
def audit(f3_objects)
  f3_objects.each do |f3_obj|
    @result = AuditResult.new(f3_obj.id, f3_obj.f3_pid, get_f4_url(f3_obj.f3_pid), nil)
    @result.status = audit_one(f3_obj)
    yield @result
  end
end

private

def audit_one(f3_obj)
  f4_obj = fedora_get_metadata(result.f4_id)
  return "Not found" if f4_obj.nil? 
  f4_model = model_from_resource f4_obj
  f3_model = normalized_f3_model(f3_obj.f3_model)
  return "Models mismatch. Expected: #{f3_model} but found: #{f4_model}" if f3_model != f4_model 
  "OK"
end

def get_f4_url pid
  @fedora_url + "/" + pid_to_uri(f3_obj.f3_pid)
end
```

mjgiarlo replied:
Just one thing to add: methods that begin with "get" or "set" are generally avoided in Ruby as a matter of convention.  Again, just a general guidelines.  If you want to get an F4 URL back, you call the method `f4_url`.

awead commented on app/models/migrate_audit_fedora4.rb:
```diff
@@ -0,0 +1,97 @@
+  def pid_to_uri(pid)
+    id = pid.split(':').last
+    raise "Cannot detect ID from PID #{pid}" unless id.length >= 9
+    id[0..1] + "/" + id[2..3] + "/" + id[4..5] + "/" + id[6..7] + "/" + id
```
Try using .join: `[ id[0..1], id[2..3], id[4..5],id[6..7], id ].join("/")`

awead commented on app/models/migrate_audit_fedora4.rb:
```diff
@@ -0,0 +1,97 @@
+  end
+
+  def model_from_resource(triples)
+    return nil if triples.nil?
```
The above correction should make this line unnecessary because we will have returned "Not Found" before we call this method.

awead commented on app/models/migrate_audit_fedora4.rb:
```diff
@@ -0,0 +1,97 @@
+  def parse_triples response
+    model = nil
+    children = []
+    start = Time.now    
```
did you put this in and forget to take it out? I can't see where you're calling it again.

awead commented on app/models/migrate_audit_fedora4.rb:
```diff
@@ -0,0 +1,97 @@
+      end
+    end
+    {model: model, children: children}
+  end
```
This is screaming refactor to me, but it's too late in the day to grok :smile:

awead commented on app/models/migrate_audit_fedora4.rb:
```diff
@@ -0,0 +1,97 @@
+  def fedora_get_metadata(url)
+    headers = { "Accept" => "application/n-triples" }
+    uri = URI.parse(url  + "/fcr:metadata")
+    response = Net::HTTP.start(uri.hostname, uri.port) { |http|
```
this is the second time you're called Net::HTTP in this commit. You could roll it out into a module but that's probably overkill. :stars:

awead commented on lib/tasks/premigrate.rake:
```diff
@@ -98,5 +103,47 @@ namespace :pre_migrate do
+  desc "Fetches PIDs for all objects in Fedora 3 for our namespace"
+  task "get_pids", [:fedora3_yml] => :environment do |cmd, args|
+    file_name = args[:fedora3_yml]
+    abort("Must specify a Fedora 3 YAML file") if file_name.nil?
```
raise insead? Never seen abort... but it could be just me.

mjgiarlo replied:
I've never seen abort either.  What's that do? :)

hectorcorrea replied:
Abort terminates the process with a returns value of false (indicating not a successful run) and sends to STDERR the error message.

awead commented on spec/models/migrate_audit_spec.rb:
```diff
@@ -0,0 +1,42 @@
+
+  before do 
+    f4_file_ok = GenericFile.create(id: "111xyzfile") { |file| file.apply_depositor_metadata('dmc') }
+    f4_file_bad = GenericFile.create(id: "333xyzfile") { |file| file.apply_depositor_metadata('dmc') }
```
I don't think you need the variables if you've not doing anything with them later. Otherwise, use `let()`

mjgiarlo replied:
:+1: to `let`. Set variables via `let` and use `before` for triggering/setting up behavior.

hectorcorrea replied:
Good catch.

awead commented on spec/models/migrate_audit_spec.rb:
```diff
@@ -0,0 +1,42 @@
+    migrated = results.select {|r| r.f3_pid == f3_file_migrated.f3_pid}.first
+    expect(migrated.status).to eq("OK")
+    expect(migrated.f4_id).to eq("#{fedora_url}/11/1x/yz/fi/111xyzfile")
+  end
```
The rspec-way of doing this is to wrap each of these it blocks in a context:
```
context "with correctly migrated files" do
  subject { results.select {|r| r.f3_pid == f3_file_migrated.f3_pid}.first }
  it "should be 'OK' do
    expect(subject.status).to eq("OK")
  end
  it "should have a valid fedora 4 url" do
    expect(subject.f4_id).to eq("#{fedora_url}/11/1x/yz/fi/111xyzfile")
  end
end
```
but this is way :eyeglasses:

mjgiarlo replied:
This is a guideline we follow inconsistently, but I prefer eschewing the word "should" in test descriptions because it's implicit.  This comes from the RSpec style guide here: http://betterspecs.org/#should  Instead I'd reword as:

```ruby
it 'returns OK' ...
it 'has a valid Fedora 4 URL' ...
```

hectorcorrea replied:
Good to know. I've updated the code.

awead commented:
@cam156 @mjgiarlo so I went :banana: :banana: :banana: with the comments. I want to say that overall, I'm :100: :+1: with this _as is_. I was just exercising a bit here, in case @hectorcorrea wanted some deeper level comments. Feel free to provide additional suggestions.

The only thing I'm not sure of are all the changes to `schema.rb` but that's just me. If we want to pull this code out eventually, that might make the schema a little weird to deal with later, but I'm not too familiar with this on the Rails side.

mjgiarlo commented on spec/models/migrate_audit_spec.rb:
```diff
@@ -0,0 +1,42 @@
+  let(:audit) { MigrateAuditFedora4.new(fedora_url, credentials[:user], credentials[:password]) }
+
+  let(:results) do 
+    results = []
```
Not necessary, but here's a little pattern I use sometimes to avoid code blocks where you set an array/hash variable, loop over something and pop stuff into the array/hash, and return the variable on a line by itself:

```ruby
[].tap do |results|
  audit.audit(f3_data) do |result|
    results.push result
  end
end
```

hectorcorrea replied:
I'll stay with the newbie approach. I am not too comfortable with (obscure) Ruby tricks just yet :)

mjgiarlo commented:
:+1: to @awead's comments! These are useful pointers. Thanks for whipping this up and working through the PR process, @hectorcorrea

hectorcorrea commented:
@awead @mjgiarlo I've implemented most of the suggestions that @awead provided. Thanks a lot for the comprehensive feedback. I also removed a few hash rockets that you guys missed ;) 

The suggestions that I didn't implement are the ones that were along the lines of "reduce these the 3 lines to a single line" when I thought the assignment of the variables line by line made the code clearer (and easier to debug) than the suggested single line approach. 

I also left a few string concatenations in place (instead of replacing them with string interpolation) when I I thought the concatenation made the operation easier to read.

mjgiarlo commented:
:clap:
